DataFormats/TrackSoA/interface/TracksSoA.h

```diff
@@ -12,89 +12,127 @@
+  // note: non-literal type 'reco::TrackLayout<128, false>::ConstViewTemplateFreeParams<128, false, true, true>::const_element'
+  // cannot be used in a constant expression
+
+  // move to use the layer gaps defined in CAGeometry
+  ALPAKA_FN_HOST_ACC inline int nLayers(const TrackSoAConstView &tracks,
+                                        const TrackHitSoAConstView &hits,
+                                        uint16_t maxLayers,
+                                        uint32_t const *__restrict__ layerStarts,
+                                        int32_t i) {
+    auto start = (i == 0) ? 0 : tracks[i - 1].hitOffsets();
+    auto end = tracks[i].hitOffsets();
+    auto hitId = hits[start].id();
+    int nl = 1;
+    auto ol = 0;
+    while (hitId >= layerStarts[ol + 1] and ol < maxLayers)
+      ++ol;
+    ++start;
+    for (; start < end; ++start) {
+      hitId = hits[start].id();
+      auto il = 0;
+      while (hitId >= layerStarts[il + 1] and il < maxLayers)
+        ++il;
+      if (il != ol)
+        ++nl;
+      ol = il;
+    }
+    return nl;
+  }
+
+  ALPAKA_FN_HOST_ACC inline float charge(const TrackSoAConstView &tracks, int32_t i) {
     //was: std::copysign(1.f, tracks[i].state()(2)). Will be constexpr with C++23
     float v = tracks[i].state()(2);
     return float((0.0f < v) - (v < 0.0f));
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float phi(ConstView const& tracks, int32_t i) {
-    return tracks[i].state()(0);
+  ALPAKA_FN_HOST_ACC inline float phi(const TrackSoAConstView &tracks, int32_t i) { return tracks[i].state()(0); }
+
+  ALPAKA_FN_HOST_ACC inline float tip(const TrackSoAConstView &tracks, int32_t i) { return tracks[i].state()(1); }
+
+  ALPAKA_FN_HOST_ACC inline float zip(const TrackSoAConstView &tracks, int32_t i) { return tracks[i].state()(4); }
+
+  ALPAKA_FN_HOST_ACC inline bool isTriplet(const TrackSoAConstView &tracks, int32_t i) {
+    return tracks[i].nLayers() == 3;
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float tip(ConstView const& tracks, int32_t i) {
-    return tracks[i].state()(1);
+  ALPAKA_FN_HOST_ACC inline int nHits(const TrackSoAConstView &tracks, int i) {
+    auto start = (i == 0) ? 0 : tracks[i - 1].hitOffsets();
+    return tracks[i].hitOffsets() - start;
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float zip(ConstView const& tracks, int32_t i) {
-    return tracks[i].state()(4);
+  // state at the beam spot: { phi, tip, 1/pt, cotan(theta), zip }
+
+  // variable of non-literal type 'MapType'
+  // (aka 'Map<Eigen::Matrix<float, 15, 1, 0, 15, 1>, 0, Eigen::InnerStride<Eigen::Dynamic>>')
+  // cannot be defined in a function before C++2b
```

Comment:
?

Reply:
I honestly don't remember what I wanted to say.

Reply:
Unless you figured out what it meant, can you drop the comment ?